src/picom.c

```diff
@@ -405,8 +405,12 @@ static void destroy_backend(session_t *ps) {
 			if (w->state == WSTATE_MAPPED) {
 				win_release_images(ps->backend_data, w);
 			} else {
+				// Unmapped windows could still have shadow images.
+				// And they might have image stale flags set by events.
 				assert(!w->win_image);
-				assert(!w->shadow_image);
+				win_clear_flags(
+				    w, WIN_FLAGS_PIXMAP_STALE | WIN_FLAGS_SHADOW_STALE);
+				win_release_images(ps->backend_data, w);
```

Comment:
Interesting, I wonder why. I think `destroy_backend` can only be reached after the critical section, so the stale flags should already be handled for mapped windows.

Reply:
- `paint_preprocess()` -> `unredirect()` -> `destroy_backend()` is called after the critical section before rendering and should be safe.
- `session_destroy()` -> `unredirect()` -> `destroy_backend()` is called in `main` after `session_run()` returns. Are we safe there?
- `handle_root_flags()` -> `configure_root()` -> `destroy_backend()` is called IN the critical section BEFORE the stale flags have been handled. This might fail if pixmaps get invalidated during a root-change (i.e. screen size changes?). I have a testcase for that, but I can't get the timing right to trigger this.

Reply:
You are right, 2 and 3 are not safe.

I don't see a good solution beside always clear the stale flags, so that's what I'll do.

Reply:
Can you post your testcase for 3 somewhere? Thanks.

Reply:
> Can you post your testcase for 3 somewhere? Thanks.

https://gist.github.com/tryone144/1103f5107448fecf5e7e02a6cf4ffb02

Reply:
@tryone144

Reply:
Hmm, looking at the code, I think IMAGES_STALE flags are only set by `win_process_update_flags`, and immediately handled later in `win_process_image_flags`; except for one case, where `map_win_start` is called in `handle_new_windows`.

That seems to be the only case where it's possible for `destroy_backend` to see a image stale flag.

Reply:
Looks like with this PR, we don't have to call `map_win_start` in `handle_new_windows` anymore. So we can eliminate case 3 (if my previous point is correct).

Reply:
@tryone144 i'm going to merge this for now.